scripts/util/doc-builder.js

```diff
@@ -0,0 +1,341 @@
+        }).then(function () {
+          return repo.getReference(self.options.docsBranchName);
+        }).then(function (reference) {
+          var docsBranchResolvedName = reference.name();
+          self.log('info', 'Blew up ', self.options.docsBranchName, ' into ', docsBranchResolvedName);
+          self.log('info', 'Getting the staging area tree id');
+          return index.writeTree().then(function (treeId) {
+            var dateObject = new Date();
+            var time = dateObject.getTime() / 1000;
+            var timeOffset = dateObject.getTimezoneOffset();
+            var author = NodeGit.Signature.create("Auto Docs",
+              "docs@shopify.com", time, timeOffset);
+            var committer = NodeGit.Signature.create("Auto Docs",
+              "docs@shopify.com", time, timeOffset);
+
+            self.log('info', 'Committing changes to \'', docsBranchResolvedName, '\'');
+            return repo.createCommit(`${docsBranchResolvedName}`, author, committer, 'Docs auto updated during build process', treeId, [docsBranchCommit]); 
+          }).then(function(commitId) {
+            self.log('info', 'New commit created: ', commitId);
+            self.log('info', 'Deleting ', self.options.versionsDirName, ' from the root directory');
+            recursiveRmdir(path.join('.', self.options.versionsDirName));
+
+            self.log('info', 'Done commiting API docs');
+            if (callback) {
+              callback(null);
+            }
+          });
+        });
+      });
+    });
+  }).catch(function (error) {
+    console.error('Error encountered while attempting to commit docs to "', self.options.docsBranchName, '"');
+    console.error(error);
+    if (callback) {
+      callback(error);
+    }
+  });
+}
+
+DocBuilder.prototype.build = function (callback) {
+  var self = this;
+
+  self.ensureDocsBranchExists(function (error) {
+    if (error) {
+      if (callback) {
+        callback(error);
+      }
+      return;
+    }
+
+    self.checkoutDocsBranch(function (error) {
+      if (error) {
+        if (callback) {
+          callback(error);
+        }
+        return;
+      }
+
+      self.checkoutAPISrc(function (error, paths) {
+        if (error) {
+          if (callback) {
+            callback(error);
+          }
+          return;
+        }
+
+        self.generateAPIDocs(JSON.parse(JSON.stringify(paths)), function () {
+
+          if (self.options.rmSrcDir) {
+            self.log('info', '\nRemoving source directories');
+            paths.forEach(function (item) {
+              var srcPath = path.join(item, self.options.srcDirName);
+              self.log('info', 'Removing ', srcPath);
+              recursiveRmdir(srcPath);
+            });
+          }
+
+          if (self.options.commitDocs) {
+            self.commitAPIDocs(callback);
+          } else if (callback) {
+            callback();
+          }
+        });
+      });
+    });
+  });
+}
```

Comment:
okay, i knew that but i was thinking more in line of the end result. Look at the below for example.
I abstract the first inner block which returns `repo.getReferences` into a function because it uses `self`. Then we abstract `repo.getReferences`'s callback into it's own function so we can `.bind()`?

This is what i have pictured in my head and that's just one part of the codebase and even worse/deep dependencies on `self` exists. Except i'm missing something @minasmart 

```
 +  NodeGit.Repository.open('.').then(function (repo) {
 +
 +    self.log('info', 'Getting available references within the repo');
 +    return repo.getReferences(NodeGit.Reference.TYPE.OID).then(function(references){
 +      var treesDataPromises = [];
 +
 +      self.log('info', 'Filtering them to include only tags and the ', self.options.masterBranchName, ' branch.');
 +      references.forEach(function(reference){
 +        var target = reference.targetPeel() || reference.target();
 +        var name = reference.shorthand();
 +        var commitPromise;
 +
 +        if(reference.isTag()) {
 +          commitPromise = repo.getCommit(target);
 +        } else if (reference.isBranch() && reference.shorthand() === self.options.masterBranchName) {
 +          commitPromise = repo.getBranchCommit(name);
 +        } else {
 +          return;
 +        }
 +        self.log('info', 'Gotten a promise for commit for ', name);
 +        self.log('info', 'Adding a promise to get commit tree for ', name);
 +        treesDataPromises.push(buildTreePromise(commitPromise, name));
 +      });
 +
 +      self.log('info', 'Resolving commit tree promises');
 +      return Promise.all(treesDataPromises);
```